File: Java/Dijkstra.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

class Pair {
    int first, second;

    public Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
}

public class Dijkstra {

    // time complexity: O((V + E) log V)
    // space complexity: O(V + E) 
    private static int[] dijkstra(List<List<Pair>> adjList, int src) {
        int n = adjList.size();
        int[] dist = new int[n + 1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[src] = 0;

        PriorityQueue<Pair> pq = new PriorityQueue<>(new Comparator<Pair>() {
            public int compare(Pair a, Pair b) {
                return a.second - b.second;
            }            
        });

        pq.add(new Pair(src, 0));

        while (!pq.isEmpty()) {
            Pair curr = pq.poll();
            int u = curr.first;
            int w = curr.second;

            if (w > dist[u]) {
                continue;
            }

            for (Pair neighbor : adjList.get(u)) {
                int v = neighbor.first;
                int weight = neighbor.second;

                if (dist[u] + weight < dist[v]) {
                    dist[v] = dist[u] + weight;
                    pq.add(new Pair(v, dist[v]));
                }
            }
        }

        return dist;
    }
    public static void main (String[] args) {
        int n, m, src;
        int[] dist;
        List<List<Pair>> adjList = new ArrayList<>();
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter the number of vertices and edges: ");
        n = sc.nextInt();
        m = sc.nextInt();
        for (int i = 0; i <= n; i++) {
            adjList.add(new ArrayList<>());
        }

        System.out.println("Enter the edges: ");
        for (int i = 0; i < m; i++) {
            int u, v, w;
            u = sc.nextInt();
            v = sc.nextInt();
            w = sc.nextInt();

            adjList.get(u).add(new Pair(v, w));
            adjList.get(v).add(new Pair(u, w));
        }

        System.out.println("Enter the source vertex: ");
        src = sc.nextInt();
        dist = dijkstra(adjList, src);

        System.out.println("Shortest distance from source vertex " + src + " to all other vertices: ");
        for (int i = 1; i <= n; i++) {
            System.out.println("Vertex " + i + ": " + dist[i]);
        }
    }
}

File: rust/dijkstra.rs
use std::cmp::Ordering;
use std::collections::BinaryHeap;
use std::io;

#[derive(Eq, PartialEq)]
struct Pair {
    vertex: usize,
    dist: i32,
}

impl Ord for Pair {
    fn cmp(&self, other: &Self) -> Ordering {
        // Reverse order for min-heap
        other.dist.cmp(&self.dist)
    }
}
impl PartialOrd for Pair {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn dijkstra(adj_list: &Vec<Vec<(usize, i32)>>, src: usize) -> Vec<i32> {
    let n = adj_list.len() - 1;
    let mut dist = vec![i32::MAX; n + 1];
    dist[src] = 0;

    let mut heap = BinaryHeap::new();
    heap.push(Pair { vertex: src, dist: 0 });

    while let Some(Pair { vertex: u, dist: w }) = heap.pop() {
        if w > dist[u] {
            continue;
        }
        for &(v, weight) in &adj_list[u] {
            if dist[u] + weight < dist[v] {
                dist[v] = dist[u] + weight;
                heap.push(Pair { vertex: v, dist: dist[v] });
            }
        }
    }
    dist
}

fn main() {
    let mut input = String::new();
    println!("Enter the number of vertices and edges: ");
    io::stdin().read_line(&mut input).unwrap();
    let nums: Vec<usize> = input.trim().split_whitespace().map(|x| x.parse().unwrap()).collect();
    let (n, m) = (nums[0], nums[1]);

    let mut adj_list = vec![vec![]; n + 1];
    println!("Enter the edges: ");
    for _ in 0..m {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let edge: Vec<i32> = input.trim().split_whitespace().map(|x| x.parse().unwrap()).collect();
        let (u, v, w) = (edge[0] as usize, edge[1] as usize, edge[2]);
        adj_list[u].push((v, w));
        adj_list[v].push((u, w));
    }

    println!("Enter the source vertex: ");
    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    let src: usize = input.trim().parse().unwrap();

    let dist = dijkstra(&adj_list, src);

    println!(
        "Shortest distance from source vertex {} to all other vertices:",
        src
    );
    for i in 1..=n {
        println!("Vertex {}: {}", i, dist[i]);
    }
}